proposals/fast-lane-ipns.md

```diff
@@ -0,0 +1,336 @@
+
+1. Name routers could cache addresses for a long time.
+2. Resolving peers could cache address locally (and require router only first time).
+3. Name routers can deny service to peers resolving name too frequently.
+4. Name routers are not involved in publishing.
+
+It is assumed that name routers will be infrastructure nodes, implying that they can be dialed from limited nodes like browsers _*(Have SSL certificates setup, or have a WebRTC address)*_ and can leverage DHT and PubSub so thay they:
+
+1. Can resolve new name on demand by raced query of DHT and PubSub.
+2. Can become aware of new names through PubSub messages.
+3. Can become aware of name keeper changes through PubSub.
+
+> Note that since name keeper changes are expected to be rare TTL can be really high.
+
+Name routing service does not has to be limited to infrastructure nodes, peer could provide that service and resolve name keeper address from a local cache.
+
+
+###### DNSLink Optimization
+
+Name owners could include addresses to chosen name keepers in the [DNSLink][] record, which would enable name resolution without resolving an address through name routers first.
+
+##### Visual illutrations
+
+###### Name resolution
+
+Diagram illustrates a name resolution flow. Note however that it does not show how name router can discover new names and their keepers from pubsub.
+
+<!--
+```flow
+pubsub=>start: PubSub
+peer=>start: 📱 Peer 
+router=>start: 🖥 Router Node.  
+keeper=>start: 🖥 Name Keeper
+resolve_keeper=>operation: Resolve QmName keeper
+resolve_name=>end: Keeper resolved
+cid=>end: QmV8
+localcached=>condition: Is keeper address cached
+routercached=>condition: Check the cache
+dht_query=>operation: Query DHT
+pubsub_query=>operation: Query PubSub
+query=>parallel: Query
+cache_router=>inputoutput: Save Keeper Address
+cache_local=>inputoutput: Save Keeper Addresss
+resolve_cid=>operation: Resolve CID for QmName
+resolved_cid=>inputoutput: QmName is QmV17
+
+peer->localcached
+  localcached(yes)->resolve_cid
+  localcached(no, right)->resolve_keeper(right)->router
+
+router(right)->routercached
+  routercached(yes)->resolve_cid
+  routercached(no, right)->query
+
+query(path1, bottom)->dht_query->cache_router
+query(path2, right)->pubsub_query->cache_router
+cache_router(left)->resolve_cid
+
+resolve_cid->keeper->resolved_cid
+pubsub->peer
+```
+-->
+![name resolution](./fast-lane-ipns/resolution.svg)
+
+
+
+###### Name update
+
+Diagram illustrating how name update flow.
+
+<--
+```flow
+owner=>start: 📱 Name Owner
+ok=>inputoutput: ✅ QmName is QmV2
+error=>inputoutput: ❌ QmV1 is not a head
+update=>operation: Update QmName from QmV1 to QmV2
+incremental=>condition: QmName is QmV1 ?
+keeper=>end: 🖥 Name Keeper
+
+
+owner->update->keeper->incremental
+incremental(yes)->ok
+incremental(no)->error
+```
+-->
+![publish](./fast-lane-ipns/publish.svg)
+
+
+###### Humane readable name resolution
+
+Diagram illustrates name resolution through DNSLink.
+
+<--
+
+```flow
```

Comment:
![image](https://user-images.githubusercontent.com/21236/109773054-9db24300-7bb3-11eb-95ed-8a5a272d32c8.png)

Reply:
I'm not sure where `a.com` is coming from.

Reply:
> I'm not sure where a.com is coming from.

As per test above this "Diagram illustrates name resolution through DNSLink", so it is an example when peer is trying to resolve `/ipns/a.com/`. @Stebalien does this make sense, otherwise could you please clarify a bit what you mean ?

Reply:
https://github.com/protocol/web3-dev-team/pull/66#discussion_r599061000